File: nim_core_v2_web/web/src/services/conversationService.ts
import {
  successRes,
  failRes,
  emit,
  formatV2ConversationLastMessage,
} from '../utils'
import { NIMResult } from '../types'
import RootService from './rootService'
import { logger } from '../logger'
import { createLoggerDecorator } from '@xkit-yx/utils'
import V2NIM from 'nim-web-sdk-ng'
import { V2NIMError } from 'nim-web-sdk-ng/dist/v2/NIM_BROWSER_SDK/types'
import {
  V2NIMConversation,
  V2NIMConversationFilter,
  V2NIMConversationOperationResult,
  V2NIMConversationOption,
  V2NIMConversationResult,
  V2NIMConversationType,
  V2NIMConversationUpdate,
} from 'nim-web-sdk-ng/dist/v2/NIM_BROWSER_SDK/V2NIMConversationService'

const TAG_NAME = 'ConversationService'
const loggerDec = createLoggerDecorator(TAG_NAME, logger)

class ConversationService {
  constructor(private rootService: RootService, private nim: V2NIM) {
    this._onConversationChanged = this._onConversationChanged.bind(this)
    this._onConversationCreated = this._onConversationCreated.bind(this)
    this._onConversationDeleted = this._onConversationDeleted.bind(this)
    this._onConversationReadTimeUpdated =
      this._onConversationReadTimeUpdated.bind(this)
    this._onSyncFailed = this._onSyncFailed.bind(this)
    this._onSyncFinished = this._onSyncFinished.bind(this)
    this._onSyncStarted = this._onSyncStarted.bind(this)
    this._onTotalUnreadCountChanged = this._onTotalUnreadCountChanged.bind(this)
    this._onUnreadCountChangedByFilter =
      this._onUnreadCountChangedByFilter.bind(this)

    nim.V2NIMConversationService.on(
      'onConversationChanged',
      this._onConversationChanged
    )
    nim.V2NIMConversationService.on(
      'onConversationCreated',
      this._onConversationCreated
    )
    nim.V2NIMConversationService.on(
      'onConversationDeleted',
      this._onConversationDeleted
    )
    nim.V2NIMConversationService.on(
      'onConversationReadTimeUpdated',
      this._onConversationReadTimeUpdated
    )
    nim.V2NIMConversationService.on('onSyncFailed', this._onSyncFailed)
    nim.V2NIMConversationService.on('onSyncFinished', this._onSyncFinished)
    nim.V2NIMConversationService.on('onSyncStarted', this._onSyncStarted)
    nim.V2NIMConversationService.on(
      'onTotalUnreadCountChanged',
      this._onTotalUnreadCountChanged
    )
    nim.V2NIMConversationService.on(
      'onUnreadCountChangedByFilter',
      this._onUnreadCountChangedByFilter
    )
  }

  destroy(): void {
    this.nim.V2NIMConversationService.off(
      'onConversationChanged',
      this._onConversationChanged
    )
    this.nim.V2NIMConversationService.off(
      'onConversationCreated',
      this._onConversationCreated
    )
    this.nim.V2NIMConversationService.off(
      'onConversationDeleted',
      this._onConversationDeleted
    )
    this.nim.V2NIMConversationService.off(
      'onConversationReadTimeUpdated',
      this._onConversationReadTimeUpdated
    )
    this.nim.V2NIMConversationService.off('onSyncFailed', this._onSyncFailed)
    this.nim.V2NIMConversationService.off(
      'onSyncFinished',
      this._onSyncFinished
    )
    this.nim.V2NIMConversationService.off('onSyncStarted', this._onSyncStarted)
    this.nim.V2NIMConversationService.off(
      'onTotalUnreadCountChanged',
      this._onTotalUnreadCountChanged
    )
    this.nim.V2NIMConversationService.off(
      'onUnreadCountChangedByFilter',
      this._onUnreadCountChangedByFilter
    )
  }

  @loggerDec
  async getConversationList(params: {
    offset: number
    limit: number
  }): Promise<NIMResult<V2NIMConversationResult>> {
    try {
      const res = await this.nim.V2NIMConversationService.getConversationList(
        params.offset,
        params.limit
      )
      return successRes({
        ...res,
        conversationList: res.conversationList.map((item) =>
          formatV2ConversationLastMessage(item)
        ),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getConversationListByOption(params: {
    option: V2NIMConversationOption
    offset: number
    limit: number
  }): Promise<NIMResult<V2NIMConversationResult>> {
    try {
      const res =
        await this.nim.V2NIMConversationService.getConversationListByOption(
          params.offset,
          params.limit,
          params.option
        )
      return successRes({
        ...res,
        conversationList: res.conversationList.map((item) =>
          formatV2ConversationLastMessage(item)
        ),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getConversation(params: {
    conversationId: string
  }): Promise<NIMResult<V2NIMConversation>> {
    try {
      return successRes(
        formatV2ConversationLastMessage(
          await this.nim.V2NIMConversationService.getConversation(
            params.conversationId
          )
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getConversationListByIds(params: {
    conversationIdList: string[]
  }): Promise<NIMResult<{ conversationList: V2NIMConversation[] }>> {
    try {
      return successRes({
        conversationList: (
          await this.nim.V2NIMConversationService.getConversationListByIds(
            params.conversationIdList
          )
        ).map((item) => formatV2ConversationLastMessage(item)),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async createConversation(params: {
    conversationId: string
  }): Promise<NIMResult<V2NIMConversation>> {
    try {
      return successRes(
        formatV2ConversationLastMessage(
          await this.nim.V2NIMConversationService.createConversation(
            params.conversationId
          )
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async deleteConversation(params: {
    conversationId: string
    clearMessage: boolean
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.deleteConversation(
          params.conversationId,
          params.clearMessage
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async deleteConversationListByIds(params: {
    conversationIdList: string[]
    clearMessage: boolean
  }): Promise<
    NIMResult<{
      conversationOperationResult: V2NIMConversationOperationResult[]
    }>
  > {
    try {
      return successRes({
        conversationOperationResult:
          await this.nim.V2NIMConversationService.deleteConversationListByIds(
            params.conversationIdList,
            params.clearMessage
          ),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async stickTopConversation(params: {
    conversationId: string
    stickTop: boolean
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.stickTopConversation(
          params.conversationId,
          params.stickTop
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async updateConversation(params: {
    conversationId: string
    updateInfo: V2NIMConversationUpdate
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.updateConversation(
          params.conversationId,
          params.updateInfo
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async updateConversationLocalExtension(params: {
    conversationId: string
    localExtension?: string
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.updateConversationLocalExtension(
          params.conversationId,
          params.localExtension || ''
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getTotalUnreadCount(): Promise<NIMResult<number>> {
    try {
      return successRes(this.nim.V2NIMConversationService.getTotalUnreadCount())
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getUnreadCountByIds(params: {
    conversationIdList: string[]
  }): Promise<NIMResult<number>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.getUnreadCountByIds(
          params.conversationIdList
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getUnreadCountByFilter(
    params: V2NIMConversationFilter
  ): Promise<NIMResult<number>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.getUnreadCountByFilter(params)
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async clearTotalUnreadCount(): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.clearTotalUnreadCount()
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async clearUnreadCountByIds(params: {
    conversationIdList: string[]
  }): Promise<
    NIMResult<{
      conversationOperationResult: V2NIMConversationOperationResult[]
    }>
  > {
    try {
      return successRes({
        conversationOperationResult:
          await this.nim.V2NIMConversationService.clearUnreadCountByIds(
            params.conversationIdList
          ),
      })
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async clearUnreadCountByGroupId(params: {
    groupId: string
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.clearUnreadCountByGroupId(
          params.groupId
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async clearUnreadCountByTypes(params: {
    conversationTypeList: V2NIMConversationType[]
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.clearUnreadCountByTypes(
          params.conversationTypeList
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async subscribeUnreadCountByFilter(params: {
    filter: V2NIMConversationFilter
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        this.nim.V2NIMConversationService.subscribeUnreadCountByFilter(
          params.filter
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async unsubscribeUnreadCountByFilter(params: {
    filter: V2NIMConversationFilter
  }): Promise<NIMResult<void>> {
    try {
      return successRes(
        this.nim.V2NIMConversationService.unsubscribeUnreadCountByFilter(
          params.filter
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async getConversationReadTime(params: {
    conversationId: string
  }): Promise<NIMResult<number>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.getConversationReadTime(
          params.conversationId
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  @loggerDec
  async markConversationRead(params: {
    conversationId: string
  }): Promise<NIMResult<number>> {
    try {
      return successRes(
        await this.nim.V2NIMConversationService.markConversationRead(
          params.conversationId
        )
      )
    } catch (error) {
      throw failRes(error as V2NIMError)
    }
  }

  private _onConversationChanged(conversationList: V2NIMConversation[]): void {
    logger.log('_onConversationChanged', conversationList)
    emit('ConversationService', 'onConversationChanged', {
      conversationList: conversationList.map((item) =>
        formatV2ConversationLastMessage(item)
      ),
    })
  }

  private _onConversationCreated(conversation: V2NIMConversation): void {
    logger.log('_onConversationCreated', conversation)
    emit(
      'ConversationService',
      'onConversationCreated',
      formatV2ConversationLastMessage(conversation)
    )
  }

  private _onConversationDeleted(conversationIds: string[]): void {
    logger.log('_onConversationDeleted', conversationIds)
    emit('ConversationService', 'onConversationDeleted', {
      conversationIdList: conversationIds,
    })
  }

  private _onConversationReadTimeUpdated(
    conversationId: string,
    readTime: number
  ): void {
    logger.log('_onConversationReadTimeUpdated', conversationId, readTime)
    emit('ConversationService', 'onConversationReadTimeUpdated', {
      conversationId,
      readTime,
    })
  }

  private _onSyncFailed(error: V2NIMError): void {
    logger.log('_onSyncFailed', error)
    emit('ConversationService', 'onSyncFailed', {})
  }

  private _onSyncFinished(): void {
    logger.log('_onSyncFinished')
    emit('ConversationService', 'onSyncFinished', {})
  }

  private _onSyncStarted(): void {
    logger.log('_onSyncStarted')
    emit('ConversationService', 'onSyncStarted', {})
  }

  private _onTotalUnreadCountChanged(unreadCount: number): void {
    logger.log('_onTotalUnreadCountChanged', unreadCount)
    emit('ConversationService', 'onTotalUnreadCountChanged', {
      unreadCount,
    })
  }

  private _onUnreadCountChangedByFilter(
    filter: V2NIMConversationFilter & {
      equals: (filter: V2NIMConversationFilter) => boolean
    },
    unreadCount: number
  ): void {
    logger.log('_onUnreadCountChangedByFilter', unreadCount, filter)
    emit('ConversationService', 'onUnreadCountChangedByFilter', {
      unreadCount,
      conversationFilter: filter,
    })
  }
}

export default ConversationService

File: nim_core_v2/lib/src/conversation/conversation_service.dart
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

// 会话服务

part of nim_core_v2;

@HawkEntryPoint()
class ConversationService {
// 消息服务
  factory ConversationService() {
    if (_singleton == null) {
      _singleton = ConversationService._();
    }
    return _singleton!;
  }

  ConversationService._();

  static ConversationService? _singleton;

  ConversationServicePlatform get _platform =>
      ConversationServicePlatform.instance;

  /// 会话同步开始
  @HawkApi(ignore: true)
  Stream<void> get onSyncStarted => _platform.onSyncStarted;

  /// 会话同步完成
  @HawkApi(ignore: true)
  Stream<void> get onSyncFinished => _platform.onSyncFinished;

  /// 会话同步失败
  @HawkApi(ignore: true)
  Stream<void> get onSyncFailed => _platform.onSyncFailed;

  /// 会话创建
  @HawkApi(ignore: true)
  Stream<NIMConversation> get onConversationCreated =>
      _platform.onConversationCreated;

  /// 会话删除
  @HawkApi(ignore: true)
  Stream<List<String>> get onConversationDeleted =>
      _platform.onConversationDeleted;

  /// 会话更新
  @HawkApi(ignore: true)
  Stream<List<NIMConversation>> get onConversationChanged =>
      _platform.onConversationChanged;

  /// 会话未读消息计数更新
  @HawkApi(ignore: true)
  Stream<int> get onTotalUnreadCountChanged =>
      _platform.onTotalUnreadCountChanged;

  /// 未读数改变回调
  @HawkApi(ignore: true)
  Stream<UnreadChangeFilterResult> get onUnreadCountChangedByFilter =>
      _platform.onUnreadCountChangedByFilter;

  /// 账号多端登录会话已读时间戳标记通知 账号A登录设备D1, D2, D1会话已读时间戳标记，同步到D2成
  @HawkApi(ignore: true)
  Stream<ReadTimeUpdateResult> get onConversationReadTimeUpdated =>
      _platform.onConversationReadTimeUpdated;

  /// 获取会话列表
  /// offset – 分页偏移，首次传0，后续拉取采用上一次返回的offset limit – 分页拉取数量，不建议超过100;
  Future<NIMResult<NIMConversationResult>> getConversationList(
      int offset, int limit) async {
    return _platform.getConversationList(offset, limit);
  }

  /// 根据查询参数获取会话列表
  /// offset – 分页偏移，首次传0，后续拉取采用上一次返回的offset
  /// limit – 分页拉取数量，不建议超过100;
  /// option – 查询选项
  Future<NIMResult<NIMConversationResult>> getConversationListByOption(
      int offset, int limit, NIMConversationOption option) async {
    return _platform.getConversationListByOption(offset, limit, option);
  }

  /// 获取会话列表，通过会话id
  /// conversationId – 会话id
  Future<NIMResult<NIMConversation>> getConversation(
      String conversationId) async {
    return _platform.getConversation(conversationId);
  }

  /// 获取会话列表，通过会话id
  /// conversationIds – 会话id列表
  Future<NIMResult<List<NIMConversation>>> getConversationListByIds(
      List<String> conversationIds) async {
    return _platform.getConversationListByIds(conversationIds);
  }

  /// 创建会话
  /// conversationId – 会话id
  Future<NIMResult<NIMConversation>> createConversation(
      String conversationId) async {
    return _platform.createConversation(conversationId);
  }

  /// 删除会话
  /// conversationId – 会话id
  /// clearMessage – 是否清除消息
  Future<NIMResult<void>> deleteConversation(
      String conversationId, bool clearMessage) async {
    return _platform.deleteConversation(conversationId, clearMessage);
  }

  /// 删除会话
  /// conversationId – 会话id
  /// clearMessage – 是否清除消息
  Future<NIMResult<List<NIMConversationOperationResult>>>
      deleteConversationListByIds(
          List<String> conversationIds, bool clearMessage) async {
    return _platform.deleteConversationListByIds(conversationIds, clearMessage);
  }

  /// 置顶会话
  /// conversationId – 会话id
  /// stickTop – 是否置顶
  Future<NIMResult<void>> stickTopConversation(
      String conversationId, bool stickTop) async {
    return _platform.stickTopConversation(conversationId, stickTop);
  }

  /// 更新会话
  /// conversationId – 会话id
  /// updateInfo – 更新信息
  Future<NIMResult<void>> updateConversation(
      String conversationId, NIMConversationUpdate updateInfo) async {
    return _platform.updateConversation(conversationId, updateInfo);
  }

  /// 更新会话本地扩展字段
  /// conversationId – 会话id
  /// localExtension – 本地扩展字段更新信息
  Future<NIMResult<void>> updateConversationLocalExtension(
      String conversationId, String localExtension) async {
    return _platform.updateConversationLocalExtension(
        conversationId, localExtension);
  }

  /// 获取会话总未读数
  Future<NIMResult<int>> getTotalUnreadCount() async {
    return _platform.getTotalUnreadCount();
  }

  /// 根据会话id获取会话未读数
  /// conversationIds – 会话id列表
  Future<NIMResult<int>> getUnreadCountByIds(
      List<String> conversationIds) async {
    return _platform.getUnreadCountByIds(conversationIds);
  }

  /// 根据过滤条件获取相应的未读数
  /// filter – 查询选项
  Future<NIMResult<int>> getUnreadCountByFilter(
      NIMConversationFilter filter) async {
    return _platform.getUnreadCountByFilter(filter);
  }

  /// 清空会话未读数
  Future<NIMResult<void>> clearTotalUnreadCount() async {
    return _platform.clearTotalUnreadCount();
  }

  /// 根据会话id清空会话未读数
  /// conversationIds – 会话id列表
  Future<NIMResult<List<NIMConversationOperationResult>>> clearUnreadCountByIds(
      List<String> conversationIds) async {
    return _platform.clearUnreadCountByIds(conversationIds);
  }

  /// 根据会话id清空会话未读数
  /// groupId – 会话分组Id
  // Future<NIMResult<void>> clearUnreadCountByGroupId(String groupId) async {
  //   return _platform.clearUnreadCountByGroupId(groupId);
  // }

  /// 根据会话类型清空相应会话的未读数
  /// groupId – 会话分组Id
  Future<NIMResult<void>> clearUnreadCountByTypes(
      List<NIMConversationType> conversationTypes) async {
    return _platform.clearUnreadCountByTypes(conversationTypes);
  }

  /// 订阅指定过滤条件的会话未读数
  /// filter – 过滤条件
  Future<NIMResult<NIMError>> subscribeUnreadCountByFilter(
      NIMConversationFilter filter) async {
    return _platform.subscribeUnreadCountByFilter(filter);
  }

  /// 取消订阅指定过滤条件的会话未读数
  /// filter – 过滤条件
  Future<NIMResult<NIMError>> unsubscribeUnreadCountByFilter(
      NIMConversationFilter filter) async {
    return _platform.unsubscribeUnreadCountByFilter(filter);
  }

  /// 获取会话已读时间戳 当前只支持P2P，高级群， 超大群
  /// conversationId – 会话id
  Future<NIMResult<int>> getConversationReadTime(String conversationId) async {
    return _platform.getConversationReadTime(conversationId);
  }

  /// 更新会话已读时间戳
  /// conversationId – 会话id
  Future<NIMResult<int>> markConversationRead(String conversationId) async {
    return _platform.markConversationRead(conversationId);
  }
}
